java-frontend/src/test/java/org/sonar/java/cfg/CFGDebug.java

```diff
@@ -17,20 +17,104 @@
  * along with this program; if not, write to the Free Software Foundation,
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
-package org.sonar.java.model;
+package org.sonar.java.cfg;
 
-import org.sonar.java.cfg.CFG;
 import org.sonar.java.cfg.CFG.Block;
+import org.sonar.java.model.SyntaxTreeDebug;
 import org.sonar.plugins.java.api.tree.Tree;
 import org.sonar.plugins.java.api.tree.Tree.Kind;
 
+import java.util.List;
+
 public class CFGDebug {
 
+  private static final String DOT_PADDING = "  ";
+  private static final String DOT_NEW_LINE = "\\n";
+
   private static final int MAX_KINDNAME = Kind.UNSIGNED_RIGHT_SHIFT_ASSIGNMENT.name().length() + 5;
 
   private CFGDebug() {
   }
 
+  public static String toDot(CFG cfg) {
+    StringBuilder sb = new StringBuilder();
+
+    List<Block> blocks = cfg.blocks();
+
+    sb.append("cfg{");
+    sb.append(DOT_NEW_LINE);
+
+    int firstBlockId = blocks.size() - 1;
+
+    for (Block block : blocks) {
+      int id = block.id();
+      sb.append(DOT_PADDING);
+      sb.append(id + dotBlockLabel(id, firstBlockId) + ";");
+      sb.append(DOT_NEW_LINE);
+    }
+
+    sb.append(DOT_NEW_LINE);
+
+    for (Block block : blocks) {
+      for (Block successor : block.successors()) {
+        sb.append(DOT_PADDING);
+        sb.append(block.id() + "->" + successor.id() + dotSuccessorLabel(block, successor) + ";");
+        sb.append(DOT_NEW_LINE);
+      }
+      for (Block exception : block.exceptions()) {
+        sb.append(DOT_PADDING);
+        sb.append(block.id() + "->" + exception.id() + dotExceptionLabel(block, exception) + ";");
+        sb.append(DOT_NEW_LINE);
+      }
+    }
+
+    sb.append("}");
+
+    return sb.toString();
+  }
+
+  private static String dotSuccessorLabel(Block block, Block successor) {
+    String edgeLabel = "";
+    if (successor == block.trueBlock()) {
+      edgeLabel = "TRUE";
+    } else if (successor == block.falseBlock()) {
+      edgeLabel = "FALSE";
+    } else if (successor == block.exitBlock()) {
+      edgeLabel = "EXIT";
+    }
+    if (!edgeLabel.isEmpty()) {
+      return "[label=\"" + edgeLabel + "\"]";
+    }
+    return "";
+  }
+
+  private static String dotExceptionLabel(Block block, Block exception) {
+    return "[label=\"EXCEPTION\",color=\"orange\",fontcolor=\"orange\"]";
+  }
+
+  private static String dotBlockLabel(int blockId, int firstBlockId) {
+    String label = "B" + blockId;
```

Comment:
This whole method could probably be refactored using StringFormat which should makes things clearer.